src/api/schema/relay.rs

```diff
@@ -0,0 +1,164 @@
+            format!("{}:{}", self.name, self.index),
+            base64::URL_SAFE_NO_PAD,
+        )
+    }
+
+    /// Decodes a base64 string into a cursor result
+    fn decode(s: &str) -> Result<Self, Base64CursorError> {
+        let bytes = base64::decode_config(s, base64::URL_SAFE_NO_PAD)
+            .map_err(Base64CursorError::DecodeError)?;
+
+        let cursor = String::from_utf8(bytes).map_err(|_| Base64CursorError::Invalid)?;
+        let index = cursor
+            .split(':')
+            .last()
+            .map(|s| s.parse::<usize>())
+            .ok_or(Base64CursorError::Invalid)?
+            .map_err(|_| Base64CursorError::Invalid)?;
+
+        Ok(Self::new(index))
+    }
+
+    /// Increment and return the index. Uses saturating_add to avoid overflow issues.
+    fn increment(&self) -> usize {
+        self.index.saturating_add(1)
+    }
+}
+
+impl From<Base64Cursor> for usize {
+    fn from(cursor: Base64Cursor) -> Self {
+        cursor.index
+    }
+}
+
+/// Makes the `Base64Cursor` compatible with Relay connections
+impl CursorType for Base64Cursor {
+    type Error = Base64CursorError;
+
+    fn decode_cursor(s: &str) -> Result<Self, Self::Error> {
+        Base64Cursor::decode(s)
+    }
+
+    fn encode_cursor(&self) -> String {
+        self.encode()
+    }
+}
+
+/// Additional fields to attach to the connection
+#[derive(SimpleObject)]
+pub struct ConnectionFields {
+    /// Total result set count
+    total_count: usize,
+}
+
+/// Relay connection result
+pub type ConnectionResult<T> = Result<Connection<Base64Cursor, T, ConnectionFields, EmptyFields>>;
+
+/// Relay-compliant connection parameters to page results by cursor/page size
+pub struct Params {
+    after: Option<String>,
+    before: Option<String>,
+    first: Option<i32>,
+    last: Option<i32>,
+}
+
+impl Params {
+    pub fn new(
+        after: Option<String>,
+        before: Option<String>,
+        first: Option<i32>,
+        last: Option<i32>,
+    ) -> Self {
+        Self {
+            after,
+            before,
+            first,
+            last,
+        }
+    }
+}
+
+/// Creates a new Relay-compliant connection. Iterator must implement `ExactSizeIterator` to
+/// determine page position in the total result set.
+pub async fn query<T, I: ExactSizeIterator<Item = T>>(
+    iter: I,
+    p: Params,
+    default_page_size: usize,
+) -> ConnectionResult<T> {
+    connection::query::<Base64Cursor, T, ConnectionFields, _, _, _>(
+        p.after,
+        p.before,
+        p.first,
+        p.last,
+        |after, before, first, last| async move {
+            let iter_len = iter.len();
+
+            let (start, end) = {
+                let after = after.map(|a| a.increment()).unwrap_or(0);
+                let before: usize = before.map(|b| b.into()).unwrap_or(iter_len);
+
+                match (first, last) {
+                    // First
+                    (Some(first), _) => (after, (after.saturating_add(first)).min(before)),
```

Comment:
Do we want to return a client error if `first` and `last` are specified? Right now it looks like we just ignore this case and default the page size.

Reply:
This is actually validated by async-graphql:

```gql
{
  components(first: 1, last: 3) {
    __typename
  }
}
```

Returns:

```json
{
  "data": null,
  "errors": [
    {
      "message": "The \"first\" and \"last\" parameters cannot exist at the same time",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "components"
      ]
    }
  ]
}
```

Reply:
Aha! Nice! Maybe a comment in this match statement would help make that invariant clearer then.

Reply:
I think ideally we'd model the function so that its inputs can only represent the valid states, but I'm imagining that might get complicated so I'm content with a comment for this case.

Reply:
👍 comment added in c56e79bb666f9359846ae05fc638c24127707ad6